rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,438 @@
+```python
+tf.copy_and_update(input_tensor, index_expression, updates)
+```
+The main difficulty is in constructing the `index_expression`.  NumPy _does_
+already have a convenience routine for this: [`np.index_exp`](https://numpy.org/doc/stable/reference/generated/numpy.s_.html),
+so the overall syntax might look like:
+```
+x = tf.copy_and_update(x, np.index_exp[:, :, 2], 0)
+```
+
+- Advantages
+  - More explicit that the tensor is being copied.
+  - More consistent with exising TensorFlow free-function syntax.
+
+- Disadvantages
+  - More verbose.
+  - `np.index_exp` is not very widely known or used.
+  - New untested API.
+  - Lack of consistency with JAX.
+  
+In discussions with the JAX team, they initially started with a syntax like this
+but then eventually switched to the current `.at` syntax due to user feedback.
+
+#### Augment Existing Tensor Scatter Methods
+
+We do already have a suite of ops that copy-and-modify the tensor:
+`tf.tensor_scatter_nd_*`.  These accept single-index and integer-tensor
+indexing. We could add new functionality to them to allow striding and boolean
+masks.  However, there are a few drawbacks:
+- The scatter/gather operations currently have a very specific meaning,
+  and have a 1-1 correspondence with dedicated ops/kernels
+  (e.g. `TensorScatterUpdate`).  If we change the meaning this will no longer
+  apply, and gather/scatter would act more like dispatchers.
+- The shape of index arguments to gather/scatter are inconsistent with the Numpy
+  convention: one is the transpose of the other.  For example,
+  ```python
+  x = np.reshape(np.arange(6), [2, 3])
+  indices = tuple([[0, 1], [0, 2]])
+  
+  y = x[indices]                               # [0, 5]
+  z1 = tf.gather_nd(x, indices)                # [1, 2]
+  z2 = tf.gather_nd(x, tf.transpose(indices))  # [0, 5]
+  ```
+  We would therefore either need to break consistency with NumPy, or break
+  backward compatibility with existing usages - neither of which is desirable.
+
+#### Mutable Tensors or tf.Variables
+
+Making existing tensors mutable is not possible nor desirable due to
+TensorFlow's current design.  We considered introducing a `MutableTensor` object
+that could modify and track an underlying tensor object.
+
+However, this would be very similar to `tf.Variable`, which does already allow
+slice assignments via the `.assign()` member function.  Instead we propose to
+allow direct assignment via `tf.Variable.__setitem__` to update a variable
+in-place.  Technically `tf.Variable` could then be used as a proxy to update a
+`tf.Tensor`:
+```python
+var_x = tf.Variable(x)
+var_x[:,:,2] = 0             # Modify the Variable's underlying contents.
+x = var_x.read_value()       # Extract the modified contents.
+```
+Forcing this to be the main update syntax for regular tensors seems a bit
+clunky and heavy-handed, since we don't need all that `tf.Variable` provides.
+The `.at[]` syntax provides a more direct option.
+  
+### Performance Implications
```

Comment:
How would this affect gradient calculation? Is `copy_and_update` going to include gradient function?

Reply:
The existing ops have gradients already.  Not much will be changed under-the-hood, everything _should_ just work.